crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,387 @@
+    fn on_forkchoice_updated(&mut self, state: ForkchoiceState) -> PayloadStatusEnum {
+        self.forkchoice_state = Some(state.clone());
+        if self.pipeline_is_idle() {
+            match self.blockchain_tree.make_canonical(&state.head_block_hash) {
+                Ok(_) => PayloadStatusEnum::Valid,
+                // TODO: handle/match error
+                Err(_error) => {
+                    self.pipeline_run_needed();
+                    PayloadStatusEnum::Syncing
+                }
+            }
+        } else {
+            PayloadStatusEnum::Syncing
+        }
+    }
+
+    /// Handle new payload message.
+    fn on_new_payload(&mut self, block: SealedBlock) -> PayloadStatusEnum {
+        if self.pipeline_is_idle() {
+            match self.blockchain_tree.insert_block(block) {
+                Ok(status) => match status {
+                    BlockStatus::Valid => PayloadStatusEnum::Valid,
+                    BlockStatus::Accepted => PayloadStatusEnum::Accepted,
+                    BlockStatus::Disconnected => PayloadStatusEnum::Syncing,
+                },
+                Err(error) => PayloadStatusEnum::Invalid { validation_error: error.to_string() },
+            }
+        } else {
+            PayloadStatusEnum::Syncing
+        }
+    }
+
+    /// Returns the next pipeline state depending on the current value of the next action.
+    /// Resets the next action to the default value.
+    fn next_pipeline_state(
+        &mut self,
+        pipeline: Pipeline<DB, U>,
+        tip: H256,
+    ) -> PipelineState<DB, U> {
+        let next_action = std::mem::take(&mut self.next_action);
+        if next_action.run_pipeline() {
+            trace!(target: "sync::controller", ?tip, "Starting the pipeline");
+            PipelineState::Running(pipeline.run_as_fut(self.db.clone(), tip))
+        } else {
+            PipelineState::Idle(pipeline)
+        }
+    }
+
+    /// Attempt to restore the tree with the finalized block number.
+    /// If the finalized block is missing from the database, trigger the pipeline run.
+    fn restore_tree_if_possible(
+        &mut self,
+        finalized_hash: BlockHash,
+    ) -> Result<(), BeaconEngineError> {
+        match self.db.view(|tx| tx.get::<tables::HeaderNumbers>(finalized_hash))?? {
+            Some(number) => self.blockchain_tree.restore_canonical_hashes(number)?,
+            None => self.pipeline_run_needed(),
+        };
+        Ok(())
+    }
+}
+
+/// On initialization, the controller will poll the message receiver and return [Poll::Pending]
+/// until the first forkchoice update message is received.
+///
+/// As soon as the controller receives the first forkchoice updated message and updates the local
+/// forkchoice state, it will launch the pipeline to sync to the head hash.
+/// While the pipeline is syncing, the controller will keep processing messages from the receiver
+/// and forwarding them to the blockchain tree.
```

Comment:
Does this mean that all new payloads etc. are sent to the blockchain tree while we sync? My worry here is that if you start sync from block 0 to e.g. 16m on mainnet that blockchain tree will grow unbounded in that period since the state in the tree is never flushed. However, this assumes that CL will send us new payloads if we are not synced yet, which might not be true.

Reply:
Yeah I would consider dropping messages as long as we're still syncing - beyond the first msg

Reply:
I would just buffer the last N.
This is needed when the pipeline is close to the tip and the transition to the tree needs to be made, buffering blocks would make the transition better.

Reply:
@onbjerg it doesn't forward **if** the pipeline is syncing